### PR TITLE
Add acceleration-based SNS IK base solver with unit test

### PR DESCRIPTION
**Pull request by robodreamer**
This commit adds acceleration-based sns ik base solver including
unit tests. It was written on top of Matt's initial code ported from 
base velocity sns ik solver.

### CODE REVIEW

File: sns_ik_lib/include/sns_ik/sns_acc_ik_base.hpp [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+class SnsAccIkBase {
+
+public:
+
+  // Smart pointer typedefs. Note: all derived classes MUST override these smart pointers.
+  typedef std::shared_ptr<SnsAccIkBase> Ptr;
+  typedef std::unique_ptr<SnsAccIkBase> uPtr;
+
+  enum class ExitCode {
```
Comment (MatthewPeterKelly) [inline sns_ik_lib/include/sns_ik/sns_acc_ik_base.hpp:44]: This duplicated the same enum that is present in the velocity solver. That is fine for now, but ultimately there should be a single base class for both solvers that removes this and other code duplication.

File: sns_ik_lib/src/sns_acc_ik_base.cpp [hunk 2]
```diff
@@ -0,0 +1,408 @@
+/** @file sns_vel_ik_base.cpp
+ *
+ * @brief The file provides the basic implementation of the SNS-IK acceleration solver
+ *
+ * @author Matthew Kelly
```
Comment (MatthewPeterKelly) [inline sns_ik_lib/src/sns_acc_ik_base.cpp]: Add @author Andy Park to the next line

File: sns_ik_lib/include/sns_ik/sns_acc_ik_base.hpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
+/** @file sns_acc_ik_base.hpp
+ *
+ * @brief The file provides the basic implementation of the SNS-IK acceleration solver
+ *
+ * @author Matthew Kelly
```
Comment (MatthewPeterKelly) [inline sns_ik_lib/include/sns_ik/sns_acc_ik_base.hpp:5]: Add @author Andy Park to the next line

File: sns_ik_lib/src/sns_acc_ik_base.cpp [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+
+SnsAccIkBase::ExitCode SnsAccIkBase::solve(const Eigen::MatrixXd& J, const Eigen::VectorXd& dJdq,
+                                           const Eigen::VectorXd& ddx, Eigen::VectorXd* ddq, double* taskScale)
+{
+  // Input validation
+  if (!ddq) { ROS_ERROR("ddq is nullptr!"); return ExitCode::BadUserInput; }
+  if (!taskScale) { ROS_ERROR("taskScale is nullptr!"); return ExitCode::BadUserInput; }
+  int nTask = ddx.size();
+  if (nTask <= 0) {
+    ROS_ERROR("Bad Input: dx.size() > 0 is required!");
```
Comment (MatthewPeterKelly) [inline sns_ik_lib/src/sns_acc_ik_base.cpp]: dx  -->  ddx

File: sns_ik_lib/test/sns_acc_ik_base_test.cpp [hunk 5]
```diff
@@ -0,0 +1,151 @@
+/**  @file sns_acc_ik_base_test.cpp
+ *
+ *  @brief Unit Test: sns_acc_ik_base solver
+ *  @author Matthew Kelly
+ *
+ *    Copyright 2018 Rethink Robotics
```
Comment (MatthewPeterKelly) [inline sns_ik_lib/test/sns_acc_ik_base_test.cpp]: Add @author Andy Park

File: sns_ik_lib/test/sns_acc_ik_base_test.cpp [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+#include <Eigen/Dense>
+#include <ros/console.h>
+
+#include <sns_ik/sns_acc_ik_base.hpp>
+#include "rng_utilities.hpp"
+
+/*************************************************************************************************/
+
+void checkEqualVector(const Eigen::VectorXd& A, const Eigen::VectorXd& B, double tol)
```
Comment (MatthewPeterKelly) [inline sns_ik_lib/test/sns_acc_ik_base_test.cpp]: This function and the next one duplicate an existing utility function in the velocity test. These should both be moved to some test utilities header that can be accessed by any test. I suggest making a new file called "test_utilities" that lives next to "rng_utilities".

File: sns_ik_lib/test/sns_acc_ik_base_test.cpp [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    ASSERT_LE(taskScale, 1.0 + tol);
+    ASSERT_GT(taskScale, 0.0);
+    checkEqualVector(taskScale * ddx, J * ddq + dJdq, tol);
+  }
+}
+
+/*************************************************************************************************/
+
+/*
+ * This test is for the solveAccIkBasic()
```
Comment (MatthewPeterKelly) [inline sns_ik_lib/test/sns_acc_ik_base_test.cpp]: Can you add some more description here?  I think the function name in the existing documentation might also be wrong.

File: sns_ik_lib/test/sns_acc_ik_base_test.cpp [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+  for (int i = 0; i < n; i++) {
+    ASSERT_LE(val(i), upp(i) + tol);
+    ASSERT_GE(val(i), low(i) - tol);
+  }
+}
+
+/*************************************************************************************************/
+
+/*
+ * This test is for the solveAccIkBasic()
```
Comment (MatthewPeterKelly) [inline sns_ik_lib/test/sns_acc_ik_base_test.cpp]: More information in the documentation about the test please.

**Review by MatthewPeterKelly — CHANGES_REQUESTED** [review]
Looks great!  I made a few suggestions, mostly related to documentation and reducing code duplication.

File: sns_ik_lib/include/sns_ik/sns_acc_ik_base.hpp (hunk 1 above)
Comment (robodreamer) [reply sns_ik_lib/include/sns_ik/sns_acc_ik_base.hpp:44]: I see. I will note that when I merge them into a single base class.

File: sns_ik_lib/test/sns_acc_ik_base_test.cpp (hunk 6 above)
Comment (robodreamer) [reply sns_ik_lib/test/sns_acc_ik_base_test.cpp]: Good suggestion!

**Review by MatthewPeterKelly — APPROVED** [review]
Excellent - ship it!

**Comment by robodreamer** [comment]
@IanTheEngineer, hey Ian if you are okay, I will go ahead and merge this pull request.

**Review by IanTheEngineer — APPROVED** [review]
This looks great, Andy! Huge improvement in additional functionality. Thanks!

**Pull request merged**